### PR TITLE
vmem: huge page support

### PR DESCRIPTION
**Pull request by stillson**
- Extended SequentialPage testing to verify huge pages
- Created general version of 4k specific calls in page table
- Took out hugepage checks
- Function to compute expected page table size if using hugepages
- fixed some typos in comments

### CODE REVIEW

File: riscv-page-tables/src/page_table.rs [hunk 1]
```diff
@@ -465,6 +460,10 @@ pub trait PagingMode {
     /// Calculates the number of PTE pages that are needed to map all pages for `num_pages` mapped
     /// pages for this type of page table.
     fn max_pte_pages(num_pages: u64) -> u64;
+
+    /// Used to calculate the number of PTE pages needed if all the pages are
+    /// not 4k
+    fn max_huge_pte_pages(num_pages: u64, page_size: PageSize) -> u64;
```
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:466]: What will `max_huge_pte_pages` be used for? If used to minimize the number of pages needed, it would limit the ability to break huge pages to 4k. That might be awkward.

File: riscv-page-tables/src/page_table.rs [hunk 2]
```diff
@@ -518,6 +518,38 @@ impl<T: PagingMode> PageTableInner<T> {
...
 
+    /// Creates a translation for `vaddr` to `paddr` with the given permissions.
+    ///Like map_4k_leaf, but with an added page size parameter.
+    ///
+    /// # Safety
+    ///
+    /// The caller must guarantee that `paddr` references a page uniquely owned by the root
+    /// `GuestStagePageTable`.
+
+    unsafe fn map_general_leaf(
```
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:529]: Is there a reason to keep `map_4k_leaf`?

File: riscv-page-tables/src/page_table.rs [hunk 3]
```diff
@@ -658,13 +698,13 @@ impl<T: FirstStagePagingMode> FirstStagePageTable<T> {
...
-        if page_size.is_huge() {
-            return Err(Error::PageSizeNotSupported(page_size));
-        }
-
         let mut inner = self.inner.lock();
-        for a in addr.iter_from().take(num_pages as usize) {
-            inner.lock_4k_leaf_for_mapping(a, get_pte_page)?;
+        for a in addr
+            .iter_from_with_size(page_size)
+            .unwrap()
```
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:704]: error instead of unwrap?

File: riscv-page-tables/src/page_table.rs (hunk 1 above)
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:466]: This and `max_pte_pages()` should probably take a length rather than a number of (assumed to be) 4kB pages.

Also, `max_huge_pte_pages()` works just fine with a non-4kB size; `max_pte_pages_with_size()` would be a better name. Then you could do something like:

```
fn max_pte_pages_with_size(len: u64, page_size: PageSize) -> u64;

fn max_pte_pages(len: u64) -> u64 {
    Self::max_pte_pages_with_size(len, PageSize::Size4k)
}
```

(BTW -- where is this function used? I don't see it called anywhere in this patch.)

File: riscv-page-tables/src/page_table.rs [hunk 4]
```diff
@@ -548,16 +580,23 @@ impl<T: PagingMode> PageTableInner<T> {
 
     /// Locks the invalid leaf PTE mapping `vaddr`, filling in any missing intermediate page tables
     /// using `get_pte_page`.
-    fn lock_4k_leaf_for_mapping(
+    /// Locks the invalid leaf PTE mapping `vaddr`, filling in any missing intermediate page tables
+    /// using `get_pte_page`.
+    fn lock_general_leaf_for_mapping(
```
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:585]: nit: drop the 'general', it's cleaner.

`lock_leaf_for_mapping`

File: riscv-page-tables/src/page_table.rs (hunk 2 above)
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:529]: This is basically identical to `map_4k_leaf()`. How about updating `map_4k_leaf()` to take a `page_size` parameter (and renaming it to just `map_leaf()`) instead?

File: riscv-page-tables/src/page_table.rs [hunk 5]
```diff
@@ -518,6 +518,38 @@ impl<T: PagingMode> PageTableInner<T> {
...
+        page_size: PageSize,
+        perms: PteFieldBits,
+    ) -> Result<()> {
+        let entry = self.walk(RawAddr::from(vaddr));
+        use TableEntryType::*;
+        match entry {
+            Locked(l) => {
+                if l.level().leaf_page_size() != page_size {
+                    // maybe not?
+                    return Err(Error::PageSizeNotSupported(l.level().leaf_page_size()));
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:542]: A `PageSizeMismatch` is probably a better fit in this case.

File: riscv-page-tables/src/page_table.rs [hunk 6]
```diff
@@ -570,12 +609,12 @@ impl<T: PagingMode> PageTableInner<T> {
             }
             Locked(_) => Err(Error::PteLocked),
             Leaf(_) => Err(Error::MappingExists),
-            Table(_) => unreachable!(),
+            Table(_) => unreachable!(), // good test of problems with page sizes
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:612]: This is case is no longer unreachable in the presence of huge pages. Consider the case where a range of address space is already mapped with 4kB pages and the user tries to map in 2MB pages. We should return an error instead of blowing up here.

File: riscv-page-tables/src/page_table.rs [hunk 7]
```diff
@@ -714,6 +754,34 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
...
+        &self,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        paddr: PageAddr<SupervisorPhys>,
+        pte_perms: PteFieldBits,
+        page_size: PageSize,
+    ) -> Result<()> {
+        let end_vaddr = self
+            .vaddr
+            .checked_add_pages_with_size(self.num_pages, page_size)
+            .unwrap();
```
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:775]: error instead of unwrap.

File: riscv-page-tables/src/page_table.rs [hunk 8]
```diff
@@ -714,6 +754,34 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
...
     }
+
+    /// Maps `vaddr` to `paddr`, The caller must guarantee that paddr points to a page it is safe to
+    /// map in this page table.
+    ///
+    /// This functions works like map_4k_addr, but takes a page size parameter.
+    /// # Safety
+    ///
+    /// Don't create aliases.
+    pub unsafe fn map_general_addr(
```
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:765]: Should `map_4k_addr` be changed to call `map_general_addr` with PageSize4k specified? Also, `map_addr` instead of `map_general_addr`.

File: riscv-page-tables/src/page_table.rs [hunk 9]
```diff
@@ -658,13 +698,13 @@ impl<T: FirstStagePagingMode> FirstStagePageTable<T> {
...
     ) -> Result<FirstStageMapper<T>> {
-        if page_size.is_huge() {
-            return Err(Error::PageSizeNotSupported(page_size));
-        }
-
         let mut inner = self.inner.lock();
-        for a in addr.iter_from().take(num_pages as usize) {
-            inner.lock_4k_leaf_for_mapping(a, get_pte_page)?;
+        for a in addr
+            .iter_from_with_size(page_size)
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:703]: Need to check that `addr` is aligned to `page_size` first.

File: riscv-pages/src/page.rs [hunk 10]
```diff
@@ -38,6 +38,22 @@ impl PageSize {
...
+            PageSize::Size4k => 12,
+            PageSize::Size2M => 21,
+            PageSize::Size1G => 30,
+            PageSize::Size512G => 39,
+        }
+    }
+
+    /// Return the number of pages of this `PageSize` that will fit in `val`.
+    /// This rounds down the number of pages, unlike num_4k_pages() which rounds up.
+    pub fn num_pages_contained_in(&self, val: u64) -> u64 {
```
Comment (dgreid) [inline riscv-pages/src/page.rs:54]: nit: would val be better names size_bytes? It took me a minute to figure out what it meant.

File: riscv-pages/src/page.rs [hunk 11]
```diff
@@ -38,6 +38,22 @@ impl PageSize {
     pub const fn num_4k_pages(val: u64) -> u64 {
         (val + PageSize::Size4k as u64 - 1) / (PageSize::Size4k as u64)
     }
+    /// Return the number of bits to shift to get one page.
+    /// i.e. 4096 == 2^12 == 1 << 12
+    pub fn page_shift(&self) -> u64 {
```
Comment (dgreid) [inline riscv-pages/src/page.rs:43]: This duplicates some code in PageTable::addr_shift, would be good to have that function use this if possible.

File: riscv-page-tables/src/page_table.rs [hunk 12]
```diff
@@ -714,6 +754,34 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
...
+            .checked_add_pages_with_size(self.num_pages, page_size)
+            .unwrap();
+
+        if vaddr < self.vaddr || vaddr >= end_vaddr {
+            return Err(Error::OutOfMapRange);
+        }
+
+        let mut inner = self.owner.inner.lock();
+
+        inner.map_general_leaf(vaddr, paddr, page_size, pte_perms)
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:783]: Need to check alignment of `vaddr` and `paddr` here (or within `map_leaf()`).

**Review by dgreid — COMMENTED** [review]
Thanks for adding the testing up front, that makes it much easier to reason about.

Couple comments.

Also it is probably worth breaking up to a few commits. any time there is a bullet list in a commit message, it's a good indication it can be broken down further.

File: riscv-page-tables/src/sv48x4.rs [hunk 13]
```diff
@@ -95,4 +95,29 @@ impl PagingMode for Sv48x4 {
         let num_l4_pages = 4;
         num_l1_pages + num_l2_pages + num_l3_pages + num_l4_pages
     }
+
+    fn max_huge_pte_pages(num_pages: u64, page_size: PageSize) -> u64 {
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/sv48x4.rs:99]: I'm having a hard time making sense of this. Is `num_pages` in units of `page_size`, or is it always 4kB? I'm skeptical of the math either way -- take for example 4MB of 4kB pages (`num_pages = 1024`). This looks like it will return 4 PTE pages, which is most definitely incorrect.

This would be a lot easier to follow if we specified the amount of address space in number of bytes (or start + end) rather than number of pages (see comment above).

File: riscv-pages/src/page.rs [hunk 14]
```diff
@@ -38,6 +38,22 @@ impl PageSize {
...
+            PageSize::Size1G => 30,
+            PageSize::Size512G => 39,
+        }
+    }
+
+    /// Return the number of pages of this `PageSize` that will fit in `val`.
+    /// This rounds down the number of pages, unlike num_4k_pages() which rounds up.
+    pub fn num_pages_contained_in(&self, val: u64) -> u64 {
+        val >> self.page_shift()
+    }
```
Comment (abrestic-rivos) [inline riscv-pages/src/page.rs:56]: Don't see this used anywhere?

File: riscv-pages/src/sequential_pages.rs [hunk 15]
```diff
@@ -39,7 +39,7 @@ pub struct SequentialPages<S: State> {
 }
```
Comment (abrestic-rivos) [inline riscv-pages/src/sequential_pages.rs:39]: I'm all for more tests, but these changes are completely unrelated from the rest of the commit that deals with mapping hugepages in a page table and so should go in their own commit.

File: riscv-page-tables/src/page_table.rs [hunk 16]
```diff
@@ -283,11 +283,6 @@ impl<'a, T: PagingMode> PageTable<'a, T> {
         }
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:283]: Commit message needs some work. Good commit messages should have prose describing the change being made and why, along with possibly highlighting any particularly tricky or notable implementation details.

File: riscv-page-tables/src/page_table.rs (hunk 2 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:529]: gone

File: riscv-page-tables/src/page_table.rs (hunk 3 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:704]: fixed

File: riscv-page-tables/src/page_table.rs (hunk 4 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:585]: fixed

File: riscv-page-tables/src/page_table.rs (hunk 8 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:765]: fixed

File: riscv-pages/src/page.rs (hunk 10 above)
Comment (stillson) [reply riscv-pages/src/page.rs:54]: fixed. It's not the right function name, but I'm not sure what to call it.

File: riscv-page-tables/src/page_table.rs (hunk 6 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:612]: fixed.

File: riscv-page-tables/src/page_table.rs (hunk 7 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:775]: fixed

File: riscv-page-tables/src/page_table.rs (hunk 1 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:466]: Got rid of the function. May come back later.

File: riscv-page-tables/src/page_table.rs (hunk 2 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:529]: done

File: riscv-page-tables/src/page_table.rs (hunk 5 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:542]: Fixed

File: riscv-page-tables/src/page_table.rs (hunk 9 above)
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:703]: fixed

File: riscv-page-tables/src/page_table.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:765]: fixed

File: riscv-page-tables/src/page_table.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (stillson) [reply riscv-page-tables/src/page_table.rs:585]: done

File: riscv-page-tables/src/page_table.rs [hunk 17]
```diff
@@ -727,10 +727,10 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
         page_size: PageSize,
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:727]: Please update the commit message. The subject is currently the same as the first patch; something like "SequentialPages: Add huge page tests" would be more appropraite.

File: riscv-page-tables/src/page_table.rs [hunk 18]
```diff
@@ -624,8 +632,7 @@ impl<T: PagingMode> PageTableInner<T> {
 
 /// A paging hierarchy for a given addressing type.
 /// This is used for HS and HU paging modes such as Sv48.
-///
-/// TODO: Support non-4k page sizes.
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:628]: Not sure this TODO is ready to come out yet -- this patch adds support for _mapping_ huge-pages, but we still need to support unmap and invalidation.

File: riscv-page-tables/src/page_table.rs [hunk 19]
```diff
@@ -658,16 +665,21 @@ impl<T: FirstStagePagingMode> FirstStagePageTable<T> {
...
+            return Err(Error::AddressMisaligned(addr.bits()));
         }
 
-        let mut inner = self.inner.lock();
-        for a in addr.iter_from().take(num_pages as usize) {
-            inner.lock_4k_leaf_for_mapping(a, get_pte_page)?;
+        let iterator = addr
+            .iter_from_with_size(page_size)
+            .ok_or_else(|| Error::AddressMisaligned(addr.bits()))?;
+
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:677]: nit: unnecessary newline

File: riscv-page-tables/src/page_table.rs [hunk 20]
```diff
@@ -658,16 +665,21 @@ impl<T: FirstStagePagingMode> FirstStagePageTable<T> {
         num_pages: u64,
         get_pte_page: &mut dyn FnMut() -> Option<Page<InternalClean>>,
     ) -> Result<FirstStageMapper<T>> {
-        if page_size.is_huge() {
-            return Err(Error::PageSizeNotSupported(page_size));
+        let mut inner = self.inner.lock();
+
+        if !addr.is_aligned(page_size) {
+            return Err(Error::AddressMisaligned(addr.bits()));
         }
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:672]: In general it's better to put these sanity checks first, before acquiring a lock, if possible.

File: riscv-page-tables/src/page_table.rs [hunk 21]
```diff
@@ -700,19 +722,33 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
...
+            return Err(Error::AddressMisaligned(vaddr.bits()));
+        }
+        if !paddr.is_aligned(self.page_size) {
+            return Err(Error::AddressMisaligned(paddr.bits()));
+        }
+
+        if let Some(end_vaddr) = self
+            .vaddr
+            .checked_add_pages_with_size(self.num_pages, self.page_size)
+        {
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:741]: This check is now redundant.

File: riscv-page-tables/src/page_table.rs (hunk 16 above)
Comment (abrestic-rivos) [reply riscv-page-tables/src/page_table.rs:289]: Getting better, though "huge page support" is still overly generic. Something like "riscv-page-tables: Add support for mapping huge pages" would be more appropriate.

File: riscv-page-tables/src/page_table.rs [hunk 22]
```diff
@@ -45,6 +49,8 @@ pub enum Error {
     OutOfMapRange,
     /// The page cannot be shared
     PageNotShareable,
+    /// Invalid address, should be page aligned
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:52]: nit: "Address is not page-size aligned" would be more clear

File: riscv-page-tables/src/page_table.rs [hunk 23]
```diff
@@ -685,13 +698,22 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
...
         num_pages: u64,
-    ) -> Self {
-        Self {
+    ) -> Result<Self> {
+        if vaddr
+            .checked_add_pages_with_size(num_pages, page_size)
+            .is_none()
+        {
+            return Err(Error::OutOfMapRange);
+        }
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:709]: Thinking about this more, there's a subtle problem with returning an error here -- it leaves the PTEs locked, effectively leaking them. I think it's probably best to do the overflow check in the caller, before we start locking PTEs.

**Review by abrestic-rivos — COMMENTED** [review]
Thanks, this is getting closer.

File: riscv-page-tables/src/page_table.rs [hunk 24]
```diff
@@ -624,7 +632,6 @@ impl<T: PagingMode> PageTableInner<T> {
 
 /// A paging hierarchy for a given addressing type.
 /// This is used for HS and HU paging modes such as Sv48.
-///
 /// TODO: Support non-4k page sizes.
```
Comment (glguida) [inline riscv-page-tables/src/page_table.rs:635]: Can this TODO be removed or there's still some non-4k pages support to be added?

Comment (glg-rv) [inline riscv-page-tables/src/page_table.rs:635]: Can this TODO be removed now or there still work to be done to support non-4k pages?

File: riscv-page-tables/src/lib.rs [hunk 25]
```diff
@@ -195,7 +195,7 @@ mod tests {
                     page.size() as usize / mem::size_of::<u64>(),
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/lib.rs:195]: you can remove the bulleted list from the commit message now

File: riscv-page-tables/src/page_table.rs (hunk 24 above)
Comment (abrestic-rivos) [reply riscv-page-tables/src/page_table.rs:635]: I actually asked for this to stay as there's quite a lot left -- we still need to support invalidating & unmapping hugepages, and then there's the whole interaction with `PageTracker`, which assumes everything is 4kB.

File: riscv-page-tables/src/page_table.rs (hunk 20 above)
Comment (abrestic-rivos) [reply riscv-page-tables/src/page_table.rs:672]: not done

File: riscv-page-tables/src/page_table.rs [hunk 26]
```diff
@@ -685,13 +703,15 @@ impl<'a, T: FirstStagePagingMode> FirstStageMapper<'a, T> {
     fn new(
         owner: &'a FirstStagePageTable<T>,
         vaddr: PageAddr<T::MappedAddressSpace>,
+        page_size: PageSize,
         num_pages: u64,
-    ) -> Self {
-        Self {
+    ) -> Result<Self> {
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:708]: no longer needs to return a Result

File: riscv-page-tables/src/page_table.rs [hunk 27]
```diff
@@ -658,16 +665,26 @@ impl<T: FirstStagePagingMode> FirstStagePageTable<T> {
...
-            inner.lock_4k_leaf_for_mapping(a, get_pte_page)?;
+        let iterator = addr
+            .iter_from_with_size(page_size)
+            .ok_or_else(|| Error::AddressMisaligned(addr.bits()))?;
+        if addr
+            .checked_add_pages_with_size(num_pages, page_size)
+            .is_none()
+        {
+            return Err(Error::OutOfMapRange);
+        }
```
Comment (abrestic-rivos) [inline riscv-page-tables/src/page_table.rs:682]: can be written more concisely as:

```
addr.checked_add_pages_with_size(num_pages, page_size).ok_or(Error::OutOfMapRnage)?;
```

File: riscv-page-tables/src/page_table.rs (hunk 24 above)
Comment (glg-rv) [reply riscv-page-tables/src/page_table.rs:635]: Yes, just realised watching the history of reviewing.

**Review by abrestic-rivos — APPROVED** [review]
